nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/AbstractEmailProcessor.java

```diff
@@ -0,0 +1,391 @@
+     */
+    protected abstract T buildMessageReceiver(ProcessContext context);
+
+    /**
+     * Return the target receivere's mail protocol (e.g., imap, pop etc.)
+     */
+    protected abstract String getProtocol(ProcessContext processContext);
+
+    /**
+     * Builds the url used to connect to the email server.
+     */
+    String buildUrl(ProcessContext processContext) {
+        String host = processContext.getProperty(HOST).evaluateAttributeExpressions().getValue();
+        String port = processContext.getProperty(PORT).evaluateAttributeExpressions().getValue();
+        String user = processContext.getProperty(USER).evaluateAttributeExpressions().getValue();
+        String password = processContext.getProperty(PASSWORD).evaluateAttributeExpressions().getValue();
+        String folder = processContext.getProperty(FOLDER).evaluateAttributeExpressions().getValue();
+
+        StringBuilder urlBuilder = new StringBuilder();
+        try {
+            urlBuilder.append(URLEncoder.encode(user, "UTF-8"));
+        } catch (UnsupportedEncodingException e) {
+            throw new ProcessException(e);
+        }
+        urlBuilder.append(":");
+        try {
+            urlBuilder.append(URLEncoder.encode(password, "UTF-8"));
+        } catch (UnsupportedEncodingException e) {
+            throw new ProcessException(e);
+        }
+        urlBuilder.append("@");
+        urlBuilder.append(host);
+        urlBuilder.append(":");
+        urlBuilder.append(port);
+        urlBuilder.append("/");
+        urlBuilder.append(folder);
+
+        String protocol = this.getProtocol(processContext);
+        String finalUrl = protocol + "://" + urlBuilder.toString();
+
+        // build display-safe URL
+        int passwordStartIndex = urlBuilder.indexOf(":") + 1;
+        int passwordEndIndex = urlBuilder.indexOf("@");
+        urlBuilder.replace(passwordStartIndex, passwordEndIndex, "[password]");
+        this.displayUrl = protocol + "://" + urlBuilder.toString();
+        if (this.logger.isInfoEnabled()) {
+            this.logger.info("Connecting to Email server at the following URL: " + this.displayUrl);
+        }
+
+        return finalUrl;
+    }
+
+    /**
+     * Builds and initializes the target message receiver if necessary (if it's
+     * null). Upon execution of this operation the receiver is fully functional
+     * and is ready to receive messages.
+     */
+    private synchronized void initializefNecessary(ProcessContext context, ProcessSession processSession) {
+        if (this.messageReceiver == null) {
+            this.processSession = processSession;
+            this.messageReceiver = this.buildMessageReceiver(context);
+
+            boolean shouldDelete = context.getProperty(SHOULD_DELETE_MESSAGES).asBoolean();
+            int fetchSize = context.getProperty(FETCH_SIZE).evaluateAttributeExpressions().asInteger();
+
+            this.messageReceiver.setShouldDeleteMessages(shouldDelete);
```

Comment:
Are you referring to the POP3 command `"DELE xxxxx"` ? If so, do not assume Trash as POP3 is notorious for lacking support of folders. Messages marked with DELE may be deleted by the server side after disconnect.

Spring seems to transparently handle that [here](https://github.com/spring-projects/spring-integration/blob/master/spring-integration-mail/src/main/java/org/springframework/integration/mail/AbstractMailReceiver.java#L409):

```
	private void postProcessFilteredMessages(Message[] filteredMessages) throws MessagingException {
		setMessageFlags(filteredMessages);

		if (shouldDeleteMessages()) {
			deleteMessages(filteredMessages);
		}
		if (this.headerMapper == null) {
			// Copy messages to cause an eager fetch
			for (int i = 0; i < filteredMessages.length; i++) {
				MimeMessage mimeMessage = new IntegrationMimeMessage((MimeMessage) filteredMessages[i]);
				filteredMessages[i] = mimeMessage;
			}
		}
	}
```

I am happy to test tomorrow morning but if Spring follows the traditional POP3 model it will perform a RETR followed by a DELE. 

That could mean the returned message exists solely within the NiFi instance retrieving it. If by any chance the processor disposeMessage method hits an issue, the message would be lost.

I will double check it tomorrow morning (AEST)